File: src/sdk/typescript/decode.ts
/**
 * Wire-object -> typed message decoding. This is the ONE place the SDK turns
 * raw JSON into the typed {@link SDKMessage} union.
 *
 * Fallback policy (so the SDK never throws on a payload it doesn't recognize):
 *   - An unknown top-level message `type` is mapped to a synthetic
 *     `SDKSystemMessage` with `subtype: "unknown"` and the raw object preserved
 *     under the original fields (via the index signature). This keeps the
 *     async iterator alive instead of crashing the caller.
 *   - An unknown content-block `type` is mapped to a `TextBlock` whose `text`
 *     is the JSON-stringified original, so the data survives in a renderable
 *     form.
 */

import type {
  ContentBlock,
  SDKAssistantMessage,
  SDKMessage,
  SDKPartialAssistantMessage,
  SDKResultMessage,
  SDKResultSubtype,
  SDKSystemMessage,
  SDKUserMessage,
  TextBlock,
  ThinkingBlock,
  ToolResultBlock,
  ToolUseBlock,
} from "./messages.js";

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function asString(value: unknown, fallback = ""): string {
  return typeof value === "string" ? value : fallback;
}

function asNumber(value: unknown, fallback = 0): number {
  return typeof value === "number" ? value : fallback;
}

function asBoolean(value: unknown, fallback = false): boolean {
  return typeof value === "boolean" ? value : fallback;
}

function decodeBlock(raw: unknown): ContentBlock {
  if (!isObject(raw)) {
    return { type: "text", text: typeof raw === "string" ? raw : JSON.stringify(raw) };
  }
  switch (raw.type) {
    case "text": {
      const block: TextBlock = { type: "text", text: asString(raw.text) };
      return block;
    }
    case "thinking": {
      const block: ThinkingBlock = {
        type: "thinking",
        thinking: asString(raw.thinking),
        signature: asString(raw.signature),
      };
      return block;
    }
    case "tool_use": {
      const block: ToolUseBlock = {
        type: "tool_use",
        id: asString(raw.id),
        name: asString(raw.name),
        input: isObject(raw.input) ? raw.input : {},
      };
      return block;
    }
    case "tool_result": {
      const content = raw.content;
      let normalized: ToolResultBlock["content"];
      if (typeof content === "string") {
        normalized = content;
      } else if (Array.isArray(content)) {
        normalized = content.filter(isObject);
      } else {
        normalized = null;
      }
      const block: ToolResultBlock = {
        type: "tool_result",
        tool_use_id: asString(raw.tool_use_id),
        content: normalized,
      };
      if (typeof raw.is_error === "boolean") {
        block.is_error = raw.is_error;
      }
      return block;
    }
    default:
      // Unknown block type: preserve the data in a renderable text block.
      return { type: "text", text: JSON.stringify(raw) };
  }
}

function decodeBlocks(raw: unknown): ContentBlock[] {
  if (!Array.isArray(raw)) {
    return [];
  }
  return raw.map(decodeBlock);
}

function decodeResultSubtype(value: unknown): SDKResultSubtype {
  if (value === "success" || value === "error_max_turns" || value === "error_during_execution") {
    return value;
  }
  return "error_during_execution";
}

/**
 * Decode a single wire message object into the typed {@link SDKMessage} union.
 * Never throws on unknown shapes — see the fallback policy at the top of this
 * file.
 */
export function decodeMessage(raw: unknown): SDKMessage {
  if (!isObject(raw)) {
    return { type: "system", subtype: "unknown", value: raw } as SDKSystemMessage;
  }

  switch (raw.type) {
    case "assistant": {
      const inner = isObject(raw.message) ? raw.message : {};
      const message: SDKAssistantMessage = {
        type: "assistant",
        message: {
          model: asString(inner.model),
          content: decodeBlocks(inner.content),
        },
        parent_tool_use_id:
          typeof raw.parent_tool_use_id === "string" ? raw.parent_tool_use_id : null,
      };
      if (typeof raw.session_id === "string") {
        message.session_id = raw.session_id;
      }
      if (typeof raw.uuid === "string") {
        message.uuid = raw.uuid;
      }
      return message;
    }

    case "user": {
      const inner = isObject(raw.message) ? raw.message : {};
      const content = typeof inner.content === "string" ? inner.content : decodeBlocks(inner.content);
      const message: SDKUserMessage = {
        type: "user",
        message: { content },
        parent_tool_use_id:
          typeof raw.parent_tool_use_id === "string" ? raw.parent_tool_use_id : null,
      };
      if (typeof raw.session_id === "string") {
        message.session_id = raw.session_id;
      }
      if (typeof raw.uuid === "string") {
        message.uuid = raw.uuid;
      }
      return message;
    }

    case "result": {
      const message: SDKResultMessage = {
        type: "result",
        subtype: decodeResultSubtype(raw.subtype),
        duration_ms: asNumber(raw.duration_ms),
        duration_api_ms: asNumber(raw.duration_api_ms),
        is_error: asBoolean(raw.is_error),
        num_turns: asNumber(raw.num_turns),
        session_id: asString(raw.session_id),
      };
      if (typeof raw.total_cost_usd === "number") {
        message.total_cost_usd = raw.total_cost_usd;
      }
      if (isObject(raw.usage)) {
        message.usage = raw.usage;
      }
      if (typeof raw.result === "string") {
        message.result = raw.result;
      }
      return message;
    }

    case "stream_event": {
      const message: SDKPartialAssistantMessage = {
        type: "stream_event",
        event: raw.event,
        session_id: asString(raw.session_id),
        parent_tool_use_id:
          typeof raw.parent_tool_use_id === "string" ? raw.parent_tool_use_id : null,
      };
      return message;
    }

    case "system": {
      // Preserve all extra fields via the index signature.
      const message = { ...raw, type: "system", subtype: asString(raw.subtype) } as SDKSystemMessage;
      return message;
    }

    default: {
      // Unknown top-level type: keep the iterator alive, preserve raw data.
      const message = { ...raw, type: "system", subtype: "unknown" } as SDKSystemMessage;
      return message;
    }
  }
}

File: src/sdk/python/decode.py
"""Wire dict -> typed Message/block decoding.

This is the ONE place wire JSON becomes typed objects. Everything else in the
package consumes already-typed values.

Unknown handling (documented choice):

* Unknown *block* types are skipped (dropped from the content list) rather than
  raising, so a newer server adding a block type does not crash older clients.
* Unknown *message* types are surfaced as a :class:`SystemMessage` with the
  original ``type`` as ``subtype`` and the full payload as ``data``. This keeps
  the run alive and the data inspectable instead of crashing or silently
  vanishing. ``stream_event`` partial deltas use this same fallback.
"""

from __future__ import annotations

from typing import Any

from blocks import (
    ContentBlock,
    TextBlock,
    ThinkingBlock,
    ToolResultBlock,
    ToolUseBlock,
)
from messages import (
    AssistantMessage,
    Message,
    ResultMessage,
    SystemMessage,
    UserMessage,
)


def decode_block(raw: dict[str, Any]) -> ContentBlock | None:
    """Decode a single content block. Returns ``None`` for unknown types."""

    block_type = raw.get("type")
    if block_type == "text":
        return TextBlock(text=raw.get("text", ""))
    if block_type == "thinking":
        return ThinkingBlock(
            thinking=raw.get("thinking", ""),
            signature=raw.get("signature", ""),
        )
    if block_type == "tool_use":
        return ToolUseBlock(
            id=raw.get("id", ""),
            name=raw.get("name", ""),
            input=raw.get("input", {}) or {},
        )
    if block_type == "tool_result":
        return ToolResultBlock(
            tool_use_id=raw.get("tool_use_id", ""),
            content=raw.get("content"),
            is_error=raw.get("is_error"),
        )
    return None


def _decode_blocks(raw_content: Any) -> list[ContentBlock]:
    if not isinstance(raw_content, list):
        return []
    blocks: list[ContentBlock] = []
    for item in raw_content:
        if isinstance(item, dict):
            block = decode_block(item)
            if block is not None:
                blocks.append(block)
    return blocks


def decode_message(raw: dict[str, Any]) -> Message:
    """Decode a wire ``message`` dict into a typed :data:`Message`.

    Unknown message types fall back to a :class:`SystemMessage`.
    """

    msg_type = raw.get("type")

    if msg_type == "assistant":
        inner = raw.get("message", {}) or {}
        return AssistantMessage(
            content=_decode_blocks(inner.get("content")),
            model=inner.get("model", ""),
            parent_tool_use_id=raw.get("parent_tool_use_id"),
        )

    if msg_type == "user":
        inner = raw.get("message", {}) or {}
        content = inner.get("content")
        decoded: str | list[ContentBlock]
        if isinstance(content, str):
            decoded = content
        else:
            decoded = _decode_blocks(content)
        return UserMessage(
            content=decoded,
            parent_tool_use_id=raw.get("parent_tool_use_id"),
        )

    if msg_type == "result":
        return ResultMessage(
            subtype=raw.get("subtype", ""),
            duration_ms=raw.get("duration_ms", 0),
            duration_api_ms=raw.get("duration_api_ms", 0),
            is_error=raw.get("is_error", False),
            num_turns=raw.get("num_turns", 0),
            session_id=raw.get("session_id", ""),
            total_cost_usd=raw.get("total_cost_usd"),
            usage=raw.get("usage"),
            result=raw.get("result"),
        )

    if msg_type == "system":
        subtype = raw.get("subtype", "")
        data = {k: v for k, v in raw.items() if k not in ("type", "subtype")}
        return SystemMessage(subtype=subtype, data=data)

    # Unknown type (including stream_event): keep the payload, do not crash.
    subtype = str(msg_type) if msg_type is not None else "unknown"
    data = {k: v for k, v in raw.items() if k != "type"}
    return SystemMessage(subtype=subtype, data=data)


def is_result(message: Message) -> bool:
    """True if the message terminates a run."""

    return isinstance(message, ResultMessage)
